src/st/no0/e_elevator.c

```diff
@@ -0,0 +1,264 @@
+    prim->u0 = prim->u2 = 0x50;
+    prim->u1 = prim->u3 = 0x60;
+    prim->x0 = prim->x2 = g_CurrentEntity->posX.i.hi - 8;
+    prim->x1 = prim->x3 = g_CurrentEntity->posX.i.hi + 8;
+    prim->v2 = prim->v3 = 0x26;
+    prim->y2 = prim->y3 = offset;
+    offset -= 0x20;
+
+    if (offset < 0x44) {
+        offset = 0x44 - offset;
+        prim->v0 = prim->v1 = offset + 6;
+        prim->y0 = prim->y1 = 0x44;
+
+        prim = prim->next;
+        prim->drawMode = DRAW_UNK02;
+        prim->v0 = prim->v1 = 0x50;
+        prim->v2 = prim->v3 = 0x60;
+        prim->y0 = prim->y1 = 0x2C;
+        prim->y2 = prim->y3 = 0x3C;
+        prim->u0 = prim->u2 = 0x22 - offset;
+        prim->u1 = prim->u3 = 0x22;
+        prim->x0 = prim->x2 = g_CurrentEntity->posX.i.hi;
+        prim->x1 = prim->x3 = g_CurrentEntity->posX.i.hi + offset;
+    } else {
+        prim->v0 = prim->v1 = 6;
+        prim->y0 = prim->y1 = offset;
+    }
+    return offset;
+}
+
+extern u16 g_EInitElevator[];
+extern u8 D_us_80180FC0[];
+extern u8 D_us_80180FD8[];
+void func_us_801C2184(Entity* self) {
+    Entity* player = &PLAYER;
+    Entity* grandparent;
+    Entity* parent;
+    Primitive* prim;
+    s32 fake = 0;
+    s32 primIndex;
+    s16 offset;
+
+    switch (self->step) {
+    case 0:
+        InitializeEntity(g_EInitElevator);
+        self->animCurFrame = 3;
+        self->zPriority = player->zPriority + 0xC;
+
+        // TODO: Define this entity in no0.h - func_us_801C26B8
+        parent = (self - 1);
+        CreateEntityFromCurrentEntity(0x29U, parent);
+        parent->params = 1;
+
+        grandparent = (self - 2);
+        CreateEntityFromCurrentEntity(0x29U, grandparent);
+        grandparent->params = 2;
+
+        primIndex = g_api.AllocPrimitives(PRIM_GT4, 12);
+        if (primIndex != -1) {
+            self->flags |= FLAG_HAS_PRIMS;
+            self->primIndex = primIndex;
+            prim = &g_PrimBuf[primIndex];
+            self->ext.prim = prim;
+            prim->tpage = 0x12;
+            prim->clut = 0x223;
+            prim->u0 = prim->u2 = 0x28;
+            prim->u1 = prim->u3 = 0x38;
+            prim->v0 = prim->v1 = 0x28;
+            prim->v2 = prim->v3 = 0x38;
+            prim->priority = 0x6B;
+            prim->drawMode = DRAW_HIDE;
+
+            prim = prim->next;
+            while (prim != NULL) {
+                prim->tpage = 0x12;
+                prim->clut = 0x223;
+                prim->priority = 0x6A;
+                prim->drawMode = DRAW_HIDE;
+                prim = prim->next;
+            }
+        } else {
+            DestroyEntity(self);
+            return;
+        }
+
+        if (player->posY.i.hi > 0xC0) {
+            self->posY.i.hi = player->posY.i.hi;
+            player->posX.i.hi = self->posX.i.hi;
+            self->animCurFrame = 0xA;
+            g_Entities[1].ext.stub[0x00] = 1;
+            SetStep(2);
+        }
+
+        break;
+    case 1:
+        if (*(u8*)&self[-1].ext.stub[0x4]) {
```

Comment:
Portability issue

Reply:
Would this make more sense updating [ET_Elevator](https://github.com/Xeeynamo/sotn-decomp/blob/16c3226a0946528a98b0dc0fb02f29a158d71d15/include/entity.h#L719) to `/* 0x80 */ u8 unk80[0x4];` and making this `if ((self-1)->ext.elevator.unk80[0]) {`?

It still matches

Reply:
Do we know what object `self[-1]` is? Are we using ET_Elevator for all the associated objects?

Reply:
self-1 should be part of the elevator (I don't know exactly what that is just yet)
https://github.com/Xeeynamo/sotn-decomp/pull/1815/files#diff-8ae1c5fc7b20f674dda925677906970cd0bc621cd02fa25cbde3414a5c344745R81-R82

It's entity 0x29 which is `func_us_801C26B8` and part of this file

Reply:
In that case using `ext.cenElevator.unk80` makes sense to me since it will have the right layout in both functions

Reply:
Done, thank you for working through it with me!